File: src/swre/swre_java.ss/src/com.saabgroup.safir.application/src/com/saabgroup/safir/application/BackdoorKeeper.java
// -*- coding: utf-8 -*-
/******************************************************************************
*
* Copyright Saab AB, 2007-2013 (http://safir.sourceforge.net)
*
* Created by: Lars Hagström / stlrha
*
*******************************************************************************
*
* This file is part of Safir SDK Core.
*
* Safir SDK Core is free software: you can redistribute it and/or modify
* it under the terms of version 3 of the GNU General Public License as
* published by the Free Software Foundation.
*
* Safir SDK Core is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Public License for more details.
*
* You should have received a copy of the GNU General Public License
* along with Safir SDK Core.  If not, see <http://www.gnu.org/licenses/>.
*
******************************************************************************/

package com.saabgroup.safir.application;


public class BackdoorKeeper
    implements  com.saabgroup.safir.dob.MessageSubscriber
{
    /**
     * Constructor.
     *
     * @param connection The connection on which the keeper will subscribe
     *                   to backdoor messages.
     */
    public BackdoorKeeper(com.saabgroup.safir.dob.ConnectionBase connection) 
    {
        if (connection == null)
        {
            throw new com.saabgroup.safir.dob.typesystem.SoftwareViolationException("You must pass a valid connection");
        }

        m_connection = connection;
    }

    /**
     * Starts subscription for backdoor commands to be sent to the Backdoor.
     *
     * The connection that was passed in the constructor must be opened before Start is called.
     *
     * If the connection is closed and opened again (maybe in a different context)
     * this method must be called again to establish the subscription.
     *
     * The class supports restarting/pausing by calling stop and then start again.
     *
     * @param backdoor [in] - Class that implements the Backdoor interface.
     * @exception Safir::Dob::NotOpenException 'Start' was called before connect to Dob.
     */
    public void start(Backdoor backdoor)
    {
        if (backdoor == null)
        {
            throw new com.saabgroup.safir.dob.typesystem.SoftwareViolationException("You must pass a valid backdoor");
        }

        stop();

        m_backdoor = backdoor;

        m_connection.subscribeMessage(com.saabgroup.safir.application.BackdoorCommand.ClassTypeId,
                                      new com.saabgroup.safir.dob.typesystem.ChannelId(),
                                      this);
        m_started = true;
    }

    public void stop()
    {
        if (!m_started)
        {
            // Never started, just return
            return; // *** RETURN ***
        }

        if (!m_connection.isOpen())
        {
            // Connection has been closed.
            return;
        }

        m_connection.unsubscribeMessage(com.saabgroup.safir.application.BackdoorCommand.ClassTypeId,
                                        com.saabgroup.safir.dob.typesystem.ChannelId.ALL_CHANNELS,
                                        this);
        m_backdoor = null;
        m_started = false;
    }

    public boolean isStarted()
    {
        return m_started;
    }

    public void onMessage(com.saabgroup.safir.dob.MessageProxy messageProxy)
    {
        com.saabgroup.safir.application.BackdoorCommand cmd =
            (com.saabgroup.safir.application.BackdoorCommand)messageProxy.getMessage();

        try
        {
            if (!cmd.nodeName().isNull())
            {
                java.util.regex.Pattern pattern = java.util.regex.Pattern.compile
                    (cmd.nodeName().getVal(),
                     java.util.regex.Pattern.CASE_INSENSITIVE);
                java.util.regex.Matcher matcher = pattern.matcher
                    (com.saabgroup.safir.dob.ThisNodeParameters.getName());
                if (!matcher.find())
                {
                    // Node name doesn't match
                    return;  // *** RETURN ***
                }
            }

            if (!cmd.connectionName().isNull())
            {
                java.util.regex.Pattern pattern = java.util.regex.Pattern.compile
                    (cmd.connectionName().getVal(),java.util.regex.Pattern.CASE_INSENSITIVE);
                java.util.regex.Matcher matcher = pattern.matcher
                    (new com.saabgroup.safir.dob.ConnectionAspectMisc(m_connection).getConnectionName());
                if (!matcher.find())
                {
                    // Connection name doesn't match
                    return;  // *** RETURN ***
                }
            }
        }
        catch (java.util.regex.PatternSyntaxException exc)
        {
            // An invalid regular expression was used, skip this command
            return;  // *** RETURN ***
        }


        if (cmd.command().isNull())
        {
            // No command given
            return;  // *** RETURN ***
        }

        String[] cmdTokens = cmd.command().getVal().split("\\s+");

        if (cmdTokens.length > 0)
        {
            if (cmdTokens[0].compareToIgnoreCase("ping") == 0)
            {
                // It's a 'ping' command. Answer it without bothering the subclass
                // implementator.
                com.saabgroup.safir.Logging.sendSystemLog
                    (com.saabgroup.safir.Logging.Severity.DEBUG,
                     "Ping reply from "
                     + new com.saabgroup.safir.dob.ConnectionAspectMisc(m_connection).getConnectionName()
                     + " on node "
                     + com.saabgroup.safir.dob.ThisNodeParameters.getName());

                return; // *** RETURN ***
            }
            else if (cmdTokens[0].compareToIgnoreCase("help") == 0)
            {
                // Get help text from subclass implementator.
                com.saabgroup.safir.Logging.sendSystemLog(com.saabgroup.safir.Logging.Severity.DEBUG,
                                                          m_backdoor.getHelpText());

                return; // *** RETURN ***
            }
        }

        // Let the subclass handle the command
        m_backdoor.handleCommand(cmdTokens);
    }

    private com.saabgroup.safir.dob.ConnectionBase m_connection;
    private Backdoor m_backdoor = null;

    private boolean m_started = false;
}

File: src/swre/swre_cpp.ss/src/BackdoorKeeper.cpp
/******************************************************************************
*
* Copyright Saab AB, 2006-2013 (http://safir.sourceforge.net)
*
* Created by: Anders Widén / stawi
*
*******************************************************************************
*
* This file is part of Safir SDK Core.
*
* Safir SDK Core is free software: you can redistribute it and/or modify
* it under the terms of version 3 of the GNU General Public License as
* published by the Free Software Foundation.
*
* Safir SDK Core is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Public License for more details.
*
* You should have received a copy of the GNU General Public License
* along with Safir SDK Core.  If not, see <http://www.gnu.org/licenses/>.
*
******************************************************************************/
#include "Safir/Application/BackdoorKeeper.h"
#include <boost/regex.hpp>

#include <Safir/Application/BackdoorCommand.h>
#include <Safir/Logging/Log.h>
#include <Safir/Dob/NodeParameters.h>
#include <Safir/Dob/ThisNodeParameters.h>
#include <Safir/Dob/ConnectionAspectMisc.h>


const Safir::Dob::Typesystem::TypeId PI_CMD_TYPE_ID = Safir::Application::BackdoorCommand::ClassTypeId;
const Safir::Dob::Typesystem::ChannelId PI_CMD_CHANNEL_ID = Safir::Dob::Typesystem::ChannelId();


namespace Safir
{
namespace Application
{

const wchar_t* delimiters = L" \t\n\r";
const wchar_t* pingCmd = L"ping";
const wchar_t* helpCmd = L"help";

//-----------------------------------------------------------------------------
BackdoorKeeper::BackdoorKeeper(const Safir::Dob::ConnectionBase&   connection)
    : m_connection(connection),
      m_backdoor(NULL),
      m_started(false)
{
}

//-----------------------------------------------------------------------------
void BackdoorKeeper::Start(Safir::Application::Backdoor& backdoor)
{
    Stop();

    m_backdoor = &backdoor;

    m_connection.SubscribeMessage(PI_CMD_TYPE_ID, PI_CMD_CHANNEL_ID, this);
    m_started = true;
}

//-----------------------------------------------------------------------------
void BackdoorKeeper::Stop()
{
    if (!m_started)
    {
        // Never started, just return
        return; // *** RETURN ***
    }

    if (!m_connection.IsOpen())
    {
        // Connection has been closed.
        return; // *** RETURN ***
    }

    m_connection.UnsubscribeMessage(PI_CMD_TYPE_ID, PI_CMD_CHANNEL_ID, this);

    m_backdoor = NULL;

    m_started = false;
}

//-----------------------------------------------------------------------------
void BackdoorKeeper::OnMessage(const Safir::Dob::MessageProxy messageProxy)
{
    try
    {
        const boost::wregex::flag_type regExpFlags = boost::regex::perl | boost::regex::icase;

        const Safir::Dob::MessagePtr message = messageProxy.GetMessage();

        const Safir::Application::BackdoorCommandPtr cmd =
                boost::dynamic_pointer_cast<Safir::Application::BackdoorCommand>(message);

        if (cmd == NULL)
        {
            // Unexpected message
            return;   // *** RETURN ***
        }

        if (!cmd->NodeName().IsNull())
        {
            if (!boost::regex_search(Safir::Dob::ThisNodeParameters::Name(),
                                     boost::wregex(cmd->NodeName().GetVal(), regExpFlags)))
            {
                // Node name doesn't match
                return;  // *** RETURN ***
            }
        }

        Safir::Dob::ConnectionAspectMisc connectionAspectMisc(m_connection);
        if (!cmd->ConnectionName().IsNull())
        {
            if (!boost::regex_search(connectionAspectMisc.GetConnectionName(),
                                     boost::wregex(cmd->ConnectionName().GetVal(), regExpFlags)))
            {
                // Connection name doesn't match
                return;  // *** RETURN ***
            }
        }

        if (cmd->Command().IsNull())
        {
            // No command given
            return;  // *** RETURN ***
        }

        // Ok, it seems that this PI-command is for this application

        std::vector<std::wstring> cmdTokens;

        Tokenize(cmd->Command().GetVal(), cmdTokens, delimiters);

        if (!cmdTokens.empty())
        {
            if (cmdTokens[0] == pingCmd)
            {
                // It's a 'ping' command. Answer to it without bothering
                // the subclass implementator.

                std::wostringstream ostr;
                ostr << "Ping reply from "
                     << connectionAspectMisc.GetConnectionName()
                     <<  " on node "
                     << Safir::Dob::ThisNodeParameters::Name();

                Safir::Logging::SendSystemLog(Safir::Logging::Debug,
                                              ostr.str());

                return; // *** RETURN ***
            }
            else if (cmdTokens[0] == helpCmd)
            {
                // Get help text from subclass implementator.
                Safir::Logging::SendSystemLog(Safir::Logging::Debug,
                                              m_backdoor->GetHelpText());

                return; // *** RETURN ***
            }
        }

        // Let the subclass handle the command
        m_backdoor->HandleCommand(cmdTokens);
    }
    catch (boost::bad_expression& /* e*/ )
    {
        // An invalid regular expression was used, skip this command
        return;  // *** RETURN ***
    }
}

//-----------------------------------------------------------------------------
void BackdoorKeeper::Tokenize(const std::wstring&        str,
                              std::vector<std::wstring>& tokens,
                              const std::wstring&        delimiters)
{
    using namespace std;

    tokens.clear();

    // Skip delimiters at beginning.
    wstring::size_type currPos = str.find_first_not_of(delimiters, 0);

    while (currPos != string::npos)
    {
        wstring::size_type lastPos = str.find_first_of(delimiters, currPos);

        wstring::size_type count;

        if (lastPos == wstring::npos)
        {
            count = lastPos;
        }
        else
        {
            count = lastPos - currPos;
        }

        tokens.push_back(str.substr(currPos, count));

        currPos = str.find_first_not_of(delimiters, lastPos);
    }
}

}
}
